pkg/rotation/reconciler.go

```diff
@@ -272,11 +272,11 @@ func (r *Reconciler) reconcile(ctx context.Context, spcps *v1alpha1.SecretProvid
 	}
 
 	providerName = string(spc.Spec.Provider)
-	providerClient, err := r.getProviderClient(providerName)
-	if err != nil {
-		errorReason = internalerrors.FailedToCreateProviderGRPCClient
-		r.generateEvent(pod, v1.EventTypeWarning, mountRotationFailedReason, fmt.Sprintf("failed to create provider client, err: %+v", err))
-		return fmt.Errorf("failed to create provider client, err: %+v", err)
+	providerClient, exists := r.providerClients[providerName]
+	if !exists {
```

Comment:
This logic is valid now as we support a mix of gRPC and old format for invoking the providers. The `--grpc-supported-providers` flag is the source of truth. When we make gRPC the default mode of communication between driver-provider, we'll need to have a fallback logic here to create the provider client on the fly is not found or we still need to continue setting the `--grpc-supported-providers` flag. WDYT?

Reply:
hmmm, good point, I need to think on that for a bit.

I'm generally a fan of explicit configuration like `--grpc-supported-providers` (I like a service's dependencies to not change during execution) but I'm also not a fan of having to edit the driver yaml for new/internal providers...

Another option would be to wrap the map in a structure with a mutex to create new clients on lookup misses. I think that would require a `--deprecated-exec-providers` flag to be added in the future.

Reply:
As a cluster operator, I can see benefits of setting the `--grpc-supported-providers` flag explicitly to control what driver-providers are allowed and rolled out in the clusters. 

But I can also see benefits of creating a provider client on the fly reduces driver downtime. 

Should we open an issue to discuss this more and get more user feedback on the desired experience?

Reply:
I've opened an issue to discuss this: https://github.com/kubernetes-sigs/secrets-store-csi-driver/issues/401.

This PR will change the current behavior where we create it on the fly. If we do go with this approach, then we should also add `azure` provider to the `--grpc-supported-providers` in all the manifests in the driver repo.

I'm not sure if users are actually piggybacking on `--grpc-supported-providers` to control which providers are used in the cluster. They would just want to configure the `SecretProviderClass` and assume the driver knows how to communicate with that provider which in the future by default will just be grpc.